### ISSUE TITLE
Unhandled Exception in demo Async (using Redis DB)

### ISSUE THREAD

**Issue by romanm-odysii** [mono/mono#11545]
## Steps to Reproduce

1. Run command "mono Async.exe" when Redis server is up and ready to receive connections.

### Current Behavior

Program crashes with "Unhandled Exception" if run with Mono.

Runs normally with .NET .

### Expected Behavior

Program shows output in console.

## On which platforms did you notice this

[ ] macOS
[*] Linux (Ubuntu 16.04)
[ ] Windows

**Version Used**: 5.16.0.179

## Stacktrace

```
Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---
```

## Source code

Source code is available at https://github.com/Plasma/booksleeve
I've removed code signing with pfx from csproj files.

**Comment by romanm-odysii** [comment]
More information:
Demo works in Mono 4.2.1.102 (Ubuntu 16.04 x64 )
Demo works in Mono 4.6.2.7 on Raspbian armv71

**Comment by lewurm** [comment]
@romanm-odysii on what platform/arch do you see this problem? x64 or armv7? could you try `export MONO_ENV_OPTIONS='-O=-aot'` and run your example code again?

**Comment by romanm-odysii** [comment]
@lewurm 
A problem described above run on Ubuntu 18.04, not 16.04. Sorry for inconvenience. 
Platform x64.
Ubuntu VM is running within VMware Player 12.5 on Windows 7 host.

netstat output:
```
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      -
```
Local instance redis-cli works OK.

Now running in Bash:
MONO_ENV_OPTIONS='-O=-aot' mono Async.exe

gives same output as before.

```
Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---
```

**Comment by romanm-odysii** [comment]
@lewurm 
A problem described above run on Ubuntu 18.04, not 16.04. Sorry for inconvenience. 
Platform x64.
Ubuntu VM is running within VMware Player 12.5 on Windows 7 host.

netstat output:
```
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      -
```
Local instance redis-cli works OK.

Now running in Bash:
MONO_ENV_OPTIONS='-O=-aot' mono Async.exe

gives same output as before.

```
Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <3833a6edf2074b959d3dab898627f0ac>:0 
  at Async.Program.Main () [0x00006] in <9299e62a45cf488581c8be38b7c30785>:0 
---> (Inner Exception #0) System.IO.IOException: The operation is not allowed on non-connected sockets.
  at System.Net.Sockets.NetworkStream.InitNetworkStream (System.Net.Sockets.Socket socket, System.IO.FileAccess Access) [0x0002a] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at System.Net.Sockets.NetworkStream..ctor (System.Net.Sockets.Socket socket) [0x00029] in <9b672a45b19f4d52b5f28f32c0c91d97>:0 
  at (wrapper remoting-invoke-with-check) System.Net.Sockets.NetworkStream..ctor(System.Net.Sockets.Socket)
  at BookSleeve.RedisConnectionBase.InitOutbound (System.Threading.Tasks.TaskCompletionSource`1[TResult] source) [0x00008] in <7ed86bdb15f24092ad2ea6704c10f39d>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Async.Program.AsyncRedisUsage () [0x000b1] in <9299e62a45cf488581c8be38b7c30785>:0 <---
```

**Comment by romiras** [comment]
@lewurm, may I assist with more information?

**Comment by lewurm** [comment]
@romiras thanks. We had a bug on our latest release on Linux/ARM, but it seems like your problem is not related to it. 

Looking again at the stack trace:
> System.AggregateException: One or more errors occurred. ---> System.IO.IOException: The operation is not allowed on non-connected sockets.

Could this be a permission problem on your operating system? e.g. port numbers <1024 on Linux require root privileges.  It's probably not because you said it works with older releases of mono, right?

@marek-safar any ideas?

**Comment by romiras** [comment]
@lewurm it doesn't looks as a permission issue. I run same program as regular user without sudo in versions 4.2/x64 and 4.6/armv71 without this issue on both of them.
Redis uses port 6379.
Ports above 1024 doesn't need root permissions in Linux, correct?

**Comment by lewurm** [comment]
yes, port 6379 does not need root permissions

**Comment by romanm-odysii** [comment]
@marek-safar,
so it is normally that program should terminate with exception in such case?

**Comment by marek-safar** [comment]
I believe so, you are using the API in not compatible way

**Comment by romanm-odysii** [comment]
@marek-safar , can you point on lines in code https://github.com/Plasma/booksleeve/blob/master/Async/Program.cs please?

**Comment by romanm-odysii** [comment]
@marek-safar, so can explain please

- why .NET app doesn't crashes with exception?
- why previous versions of Mono didn't crashed?

**Comment by romanm-odysii** [comment]
@marek-safar and @lewurm ,
it appears a bug was resolved in latest update in 5.18.0.225.
At least it works now in both Ubuntu 16.04 and 18.04.

**Comment by lewurm** [comment]
Thanks for reporting that. I'll close this issue. Feel free to reopen if you have further problems.
